OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -39,6 +43,203 @@ static double UTOL = (double)SimTK::Eps*1e2;
 static double INTTOL = (double)SimTK::Eps*1e2;
 static int MAXITER = 20;
 static int NUM_SAMPLE_PTS = 100;
+
+//=============================================================================
+// PARAMETERS
+//=============================================================================
+
+/** Helper struct containing all parameters required to construct the
+ * SmoothSegmentedFunctionData.
+*/
+namespace OpenSim {
+struct SmoothSegmentedFunctionParameters {
+
+    SmoothSegmentedFunctionParameters(const SimTK::Matrix& mX,
+            const SimTK::Matrix& mY, double x0, double x1, double y0, double y1,
+            double dydx0, double dydx1, bool computeIntegral, bool intx0x1)
+            : _mX(mX), _mY(mY), _x0(x0), _x1(x1), _y0(y0), _y1(y1),
+              _dydx0(dydx0), _dydx1(dydx1), _computeIntegral(computeIntegral),
+              _intx0x1(intx0x1) {}
+
+    SimTK::Matrix _mX;
+    SimTK::Matrix _mY;
+
+    double _x0;
+    double _x1;
+    double _y0;
+    double _y1;
+    double _dydx0;
+    double _dydx1;
+    bool _computeIntegral;
+    bool _intx0x1;
+};
+
+bool operator==(
+    const SimTK::Matrix &lhs,
+    const SimTK::Matrix &rhs
+) {
+    if (lhs.nrow() != rhs.nrow() || lhs.ncol() != rhs.ncol()) {
+        return false;
+    }
+    for (int r = 0; r < lhs.nrow(); ++r)
+        for (int c = 0; c < lhs.ncol(); ++c)
+            if (lhs.row(r)[c] != rhs.row(r)[c]) return false;
+    return true;
+}
+
+bool operator==(
+    const SmoothSegmentedFunctionParameters &lhs,
+    const SmoothSegmentedFunctionParameters &rhs
+) {
+    return
+    lhs._x0 == rhs._x0 &&
+    lhs._x1 == rhs._x1 &&
+    lhs._y0 == rhs._y0 &&
+    lhs._y1 == rhs._y1 &&
+    lhs._dydx0 == rhs._dydx0 &&
+    lhs._dydx1 == rhs._dydx1 &&
+    lhs._computeIntegral == rhs._computeIntegral &&
+    lhs._intx0x1 == rhs._intx0x1 &&
+    lhs._mX == rhs._mX &&
+    lhs._mY == rhs._mY;
+}
+
+//=============================================================================
+// HASHING OF PARAMETERS
+//=============================================================================
+
+template<typename T>
+inline std::size_t HashCombine(std::size_t seed, T const &v) {
+    std::hash<T> hasher;
+    return seed ^ (hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2));
+}
+
+template<typename T>
+inline std::size_t HashOf(T const &v) {
+    return std::hash<T>{}(v);
+}
+
+template<typename T, typename... Others>
+inline std::size_t HashOf(T const &v, Others const &...others) {
+    return HashCombine(HashOf(v), HashOf(others...));
+}
+
+template<>
+inline std::size_t HashOf(SimTK::Matrix const &matrix) {
+    std::size_t hash = HashOf(matrix.nrow(), matrix.ncol());
```

Comment:
You don't need to seed this hash with the number of rows and the number of columns: iterating over the contents and combining that with a seed should be enough (or returning the seed as-is if the matrix is empty).

Reply:
My reasoning was, that because the matrix is hashed first, there is no seed yet. Also, it should be possible to compute the hash of just a matrix, so if it is empty I should return something. I could create a default hash in that case, but this seemed like a good alternative. But perhaps I misunderstand.